// discord.js-commando/src/commands/util/help.js
const { stripIndents, oneLine } = require('common-tags');
const Command = require('../base');
const { disambiguation } = require('../../util');

module.exports = class HelpCommand extends Command {
	constructor(client) {
		super(client, {
			name: 'help',
			group: 'util',
			memberName: 'help',
			aliases: ['commands'],
			description: 'C\'est ce que tu regardes actuellement ! Te permet de voir la liste des commandes existantes !',
			examples: ['help', 'help prefix'],
			guarded: true,

			args: [
				{
					key: 'command',
					prompt: 'Pour quelle commande veux tu des informations ?',
					type: 'string',
					default: ''
				}
			]
		});
	}

	async run(msg, args) { // eslint-disable-line complexity
		const groups = this.client.registry.groups;
		const commands = this.client.registry.findCommands(args.command, false, msg);
		const showAll = args.command && args.command.toLowerCase() === 'all';
		if(args.command && !showAll) {
			if(commands.length === 1) {
				let help = stripIndents`
					${oneLine`
						__Command **${commands[0].name}**:__ ${commands[0].description}
						${commands[0].guildOnly ? ' (Usable only in servers)' : ''}
						${commands[0].nsfw ? ' (NSFW)' : ''}
					`}

					**Format:** ${msg.anyUsage(`${commands[0].name}${commands[0].format ? ` ${commands[0].format}` : ''}`)}
				`;
				if(commands[0].aliases.length > 0) help += `\n**Aliases:** ${commands[0].aliases.join(', ')}`;
				help += `\n${oneLine`
					**Group:** ${commands[0].group.name}
					(\`${commands[0].groupID}:${commands[0].memberName}\`)
				`}`;
				if(commands[0].details) help += `\n**Details:** ${commands[0].details}`;
				if(commands[0].examples) help += `\n**Examples:**\n${commands[0].examples.join('\n')}`;

				const messages = [];
				try {
					messages.push(await msg.direct(help));
					if(msg.channel.type !== 'dm') messages.push(await msg.reply('Je t\'ai envoyé un DM ! :blush:'));
				} catch(err) {
					messages.push(await msg.reply('Je n\'ai pas pu t\'envoyer de DM ! Verifies que tu les as activés !'));
				}
				return messages;
			} else if(commands.length > 15) {
				return msg.reply('Il y a plusieurs commandes trouvées ! Tu peux être plus spécifique, s\'il te plait ?');
			} else if(commands.length > 1) {
				return msg.reply(disambiguation(commands, 'commands'));
			} else {
				return msg.reply(
					`Impossible d'identifier cette commande ! Utilises ${msg.usage(
						null, msg.channel.type === 'dm' ? null : undefined, msg.channel.type === 'dm' ? null : undefined
					)} pour voir la liste des commandes !`
				);
			}
		} else {
			const messages = [];
			try {
				messages.push(await msg.direct(stripIndents`
					${oneLine`
						Pour faire une commande dans ${msg.guild ? msg.guild.name : 'n\'importe quel serveur'},
						utilises ${Command.usage('command', msg.guild ? msg.guild.commandPrefix : null, this.client.user)}.
						Par exemple, ${Command.usage('prefix', msg.guild ? msg.guild.commandPrefix : null, this.client.user)}.
					`}
					Pour lancer une commande dans cette discussion, utilises simplement ${Command.usage('command', null, null)} sans préfixe.

					Utilises ${this.usage('<command>', null, null)} pour voir les informations détaillées sur une commande.
					Utilises ${this.usage('all', null, null)} pour voir la liste de toutes les commandes, pas seulement celles disponibles.

					__**${showAll ? 'Toutes les commandes' : `Commandes disponibles dans ${msg.guild || 'cette conversation'}`}**__

					${(showAll ? groups : groups.filter(grp => grp.commands.some(cmd => cmd.isUsable(msg))))
						.map(grp => stripIndents`
							__${grp.name}__
							${(showAll ? grp.commands : grp.commands.filter(cmd => cmd.isUsable(msg)))
								.map(cmd => `**${cmd.name}:** ${cmd.description}${cmd.nsfw ? ' (NSFW)' : ''}`).join('\n')
							}
						`).join('\n\n')
					}
				`, { split: true }));
				if(msg.channel.type !== 'dm') messages.push(await msg.reply('Je t\'ai envoyé un DM ! :blush:'));
			} catch(err) {
				messages.push(await msg.reply('Je n\'ai pas pu t\'envoyer de DM ! Verifies que tu les as activés !'));
			}
			return messages;
		}
	}
};
