// discord.js-commando/src/commands/util/ping.js
const { oneLine } = require('common-tags');
const Command = require('../base');

module.exports = class PingCommand extends Command {
	constructor(client) {
		super(client, {
			name: 'ping',
			group: 'util',
			memberName: 'ping',
			description: 'Permet de vérifier ma réactivité dans le serveur !',
			throttling: {
				usages: 5,
				duration: 10
			}
		});
	}

	async run(msg) {
		if(!msg.editable) {
			const pingMsg = await msg.reply('Ping en cours...');
			return pingMsg.edit(oneLine`
				${msg.channel.type !== 'dm' ? `${msg.author},` : ''}
				Pong! L'opération a pris ${pingMsg.createdTimestamp - msg.createdTimestamp}ms.
				${this.client.ping ? `Le Heartbeat Ping a pris ${Math.round(this.client.ping)}ms.` : ''}
			`);
		} else {
			await msg.edit('Ping en cours...');
			return msg.edit(oneLine`
				Pong! L'opération a pris ${msg.editedTimestamp - msg.createdTimestamp}ms.
				${this.client.ping ? `Le Heartbeat Ping a pris ${Math.round(this.client.ping)}ms.` : ''}
			`);
		}
	}
};
